### PR TITLE
[Prompty] Support model config in prompty

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
- prompty with azure openai
```
---
name: Basic Prompt
description: A basic prompt that uses the GPT-3 chat API to answer questions
model:
    api: chat
    configuration:
      type: azure_openai
      azure_deployment: gpt-35-turbo
      azure_endpoint: ${env:AZURE_ENDPOINT}
      api_key: ${env:AZURE_API_KEY}
    parameters:
      max_tokens: 128
      temperature: 0.2
inputs:
  firstName:
    type: string
    default: John
  lastName:
    type: string
    default: Doh
  question:
    type: string
---
system:
You are an AI assistant who helps people find information.
Use their name to address them in your responses.

user:
{{question}}

```
- prompty with connection
```
---
name: Basic Prompt
description: A basic prompt that uses the GPT-3 chat API to answer questions
model:
    api: chat
    connection: azure_open_ai_connection
    configuration:
      type: azure_openai
      azure_deployment: gpt-35-turbo
    parameters:
      max_tokens: 128
      temperature: 0.2
inputs:
  firstName:
    type: string
    default: John
  lastName:
    type: string
    default: Doh
  question:
    type: string
---
system:
You are an AI assistant who helps people find information.
Use their name to address them in your responses.

user:
{{question}}

```

- prompty with openai
```
---
name: Basic Prompt
description: A basic prompt that uses the GPT-3 chat API to answer questions
model:
    api: chat
    connection: azure_open_ai_connection
    configuration:
      type: openai
      model: gpt-35-turbo
      api_key: ${env:API_KEY}
      base_url: ${env:BASE_URL}
    parameters:
      max_tokens: 128
      temperature: 0.2
inputs:
  firstName:
    type: string
    default: John
  lastName:
    type: string
    default: Doh
  question:
    type: string
---
system:
You are an AI assistant who helps people find information.
Use their name to address them in your responses.

user:
{{question}}

```

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 1]
```diff
@@ -15,6 +16,41 @@
 from promptflow.core._utils import render_jinja_template_content
 
 
+@dataclass
+class PromptyModelConfiguration:
+    api: str
+    configuration: dict
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:22]: are we plan to add the AOAIModelConfig class in next PR?

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 2]
```diff
@@ -15,6 +16,41 @@
 from promptflow.core._utils import render_jinja_template_content
 
 
+@dataclass
+class PromptyModelConfiguration:
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:20]: move config class to another file

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 3]
```diff
@@ -15,6 +16,41 @@
...
+    api: str
+    configuration: dict
+    parameters: dict
+    response: str = "first"
+    connection: InitVar[Union[str, _Connection]] = None
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:25]: I think the connection will be in configuration dict following @D-W- 's design

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 2 above)
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:20]: add some docstring to this class

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 4]
```diff
@@ -15,6 +16,41 @@
...
+        if connection:
+            self._connection_config = connection
+        else:
+            self._connection_config = copy.copy(self.configuration)
+            if self._connection_config.get("type", None) == "azure_openai":
+                self._connection_config["api_base"] = self._connection_config.get("azure_endpoint", None)
+
+    def get_connection(self):
+        if not hasattr(self, "_connection"):
+            self._connection = get_connection(self._connection_config)
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:37]: the resolve logic should postpone to executor

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: add AzureOpenAIModelConfiguration and OpenAIModelConfiguration

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: move to _model_configuration.py

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: move connection to configuration dict

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: add doc string

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: fix it

File: src/promptflow-core/promptflow/core/_model_configuration.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    pass
+
+
+@dataclass
+class AzureOpenAIModelConfiguration(ModelConfiguration):
+    azure_deployment: str
+    azure_endpoint: str = None
+    api_version: str = None
+    api_key: str = None
+    azure_ad_token_provider: Callable = None  # AzureADTokenProvider is a callable object to retrieve token
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_model_configuration.py:18]: I think we should not support this?

Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_model_configuration.py]: remove it

**Review by wangchao1230 — APPROVED** [review]

**Review by 0mza987 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**